tempesta_fw/t/unit/sse_parser/sse_parser.c

```diff
@@ -0,0 +1,1552 @@
+    __m128i directoryPatterns;
+    //bitmasks for different types of data
+    __m128i alnumDotMinusBitmask;
+    __m128i hostnameLiteralBitmask;
+    __m128i uriBitmask;
+    __m128i uriArgsBitmask;
+    __m128i uriAsIsBitmask;
+    __m128i headerNameBitmask;
+    __m128i headerValueBitmask;
+
+    union {
+        TokenSet ts;
+        __m128i  buffer[25];
+    } http_method;
+    union {
+        TokenSet ts;
+        __m128i  buffer[9];
+    } http_schema;
+    union {
+        TokenSet ts;
+        __m128i  buffer[17];
+    } http_version;
+};
+
+static struct Constants cc;
+
+void sse_init_constants() {
+    struct Constants * c = &cc;
+    int ret;
+
+    c->shuffle[0] = c->shuffle[1] = _mm_setr_epi8(
+                0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15);
+    c->shuffle[2] = _mm_set1_epi8(0xFF);
+    c->shuffle[3] = _mm_setzero_si128();
+    c->masktest1  = _mm_setr_epi8(1,2,4,8,16,32,64,128,0,0,0,0,0,0,0,0);
+    c->masktest2  = _mm_set1_epi8(0xF);
+    c->tokencmp   = _mm_setr_epi8(1, 0, 2, 0, 4, 0, 8, 0, 16, 0, 32, 0, 64, 0, 128, 0);
+    c->swapbytes16 = _mm_set_epi8(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15);
+    c->cvt10_2_bytes = _mm_set1_epi16(256+10);
+    c->cvt10_2_words = _mm_set1_epi32(65536+100);
+
+    c->spaces = _mm_set1_epi8(0x20);
+    c->semicolons = _mm_set1_epi8(':');
+    c->slashes = _mm_set1_epi8('/');
+    c->tabs = _mm_set1_epi8('\t');
+    c->plus = _mm_set1_epi8('+');
+    c->nine = _mm_set1_epi8('9');
+    c->zero = _mm_set1_epi8('0');
+    c->_n   = _mm_set1_epi8('\n');
+    c->_r   = _mm_set1_epi8('\r');
+
+    c->tabs2spaces = _mm_set1_epi8(' '-'\t');
+    c->plus2spaces = _mm_set1_epi8('+'-' ');
+
+    c->masktest1 = _mm_setr_epi8(1,2,4,8,16,32,64,128,0,0,0,0,0,0,0,0);
+    c->masktest2 = _mm_set1_epi8(0xF);
+
+    c->newline1 = _mm_setr_epi8('\n','\n', 0  , 0  ,
+                                '\r','\n','\n', 0  ,
+                                '\n','\r','\n', 0  ,
+                                '\r','\n','\r','\n');
+    c->newline2 = _mm_setr_epi8(0xFF,0xFF, 0  , 0  ,
+                                0xFF,0xFF,0xFF, 0  ,
+                                0xFF,0xFF,0xFF, 0  ,
+                                0xFF,0xFF,0xFF,0xFF);
+    c->dq0 = _mm_setr_epi8(
+                0x01, 0x01, 0x01, 0x01, 0x02, 0x02, 0x02, 0x02,
+                0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF);
+    c->dq1 = _mm_setr_epi8(
+                0x00, 0x00, 0x20, 0x20, 0x00, 0x00, 0x20, 0x20,
+                0x00, 0x00, 0x20, 0x20, 0x00, 0x00, 0x20, 0x20);
+    c->dq2 = _mm_setr_epi8(
+                0x2F, 0x39, 0x60, 0x66, 0x2F, 0x39, 0x60, 0x66,
+                0x2F, 0x39, 0x60, 0x66, 0x2F, 0x39, 0x60, 0x66);
+    c->dq3 = _mm_set1_epi16(0x00FF);
+    c->dq4 = _mm_setr_epi8(
+                0x30, 0xFF, 0x57, 0xFF, 0x30, 0xFF, 0x57, 0xFF,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x07, 0x00);
+    c->dq5 = _mm_setr_epi16(
+                16, 16, 1, 1, 0, 0, 0, 0);
```

Comment:
These constants are for SSE decoder of %xx sequences(%20 for example)

dq0 constant expands 1st and 2nd byte in 4 positions each.

we need 4 copies to perform 4 comparisons c>='0', c<='9', c>='a', c<='f'

dq1 is to lowercase positions which may be  [A-F]

dq2 is actual comparison mask ('0'-1, '9'+1, 'a'-1, 'f'+1)

if comparison was ok, we will get a 00FF word (dq3 is to ensure we found it)

dq4 is used to convert value obtained by dq0-dq1 and masking with result of
dq2-dq3 to decimal values

dq5 is used to combine upper and lower hex symbols into a hex code

2016-06-14 14:17 GMT+04:00 Alexander Krizhanovsky <notifications@github.com>
:

> In tempesta_fw/t/unit/sse_parser/sse_parser.c
> <https://github.com/tempesta-tech/tempesta/pull/550#discussion_r66944236>:
>
> > +                                0xFF,0xFF,0xFF,0xFF);
> > +    c->dq0 = _mm_setr_epi8(
> > +                0x01, 0x01, 0x01, 0x01, 0x02, 0x02, 0x02, 0x02,
> > +                0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF);
> > +    c->dq1 = _mm_setr_epi8(
> > +                0x00, 0x00, 0x20, 0x20, 0x00, 0x00, 0x20, 0x20,
> > +                0x00, 0x00, 0x20, 0x20, 0x00, 0x00, 0x20, 0x20);
> > +    c->dq2 = _mm_setr_epi8(
> > +                0x2F, 0x39, 0x60, 0x66, 0x2F, 0x39, 0x60, 0x66,
> > +                0x2F, 0x39, 0x60, 0x66, 0x2F, 0x39, 0x60, 0x66);
> > +    c->dq3 = _mm_set1_epi16(0x00FF);
> > +    c->dq4 = _mm_setr_epi8(
> > +                0x30, 0xFF, 0x57, 0xFF, 0x30, 0xFF, 0x57, 0xFF,
> > +                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x07, 0x00);
> > +    c->dq5 = _mm_setr_epi16(
> > +                16, 16, 1, 1, 0, 0, 0, 0);
>
> What the constants mean and what do they have to do with quotes?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tempesta-tech/tempesta/pull/550/files/939d52c95d17fae07df56bec93d27a25c376571b#r66944236>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABnrSVWCa6eX8_Gp6WDTIWcftU8UuB8Fks5qLn_XgaJpZM4IssRS>
> .
>